### PR TITLE
improve pipeline run fetcher

### PR DESCRIPTION
**Pull request by rinckm**
The controller now uses a pipeline run fetcher based on the PipelineRunLister. 
This makes use of the cache and should be much faster.

### CODE REVIEW

File: pkg/k8s/pipelineRunFetcher.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+// NewPipelineRunListerFetcher returns an operative implementation of PipelineRunFetcher
+func NewPipelineRunListerFetcher(lister stewardLister.PipelineRunLister) PipelineRunFetcher {
+	return &pipelineRunListerFetcher{
+		lister: lister,
+	}
+}
+
+// ByName fetches PipelineRun resource from Kubernetes by name and namespace
+// Return nil,nil if specified pipeline does not exist
```
Comment (anbrsap) [inline pkg/k8s/pipelineRunFetcher.go]: Put the docs on the interface to make them visible there. Here you just need to refer to the interface.

File: pkg/k8s/pipelineRunFetcher.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+// PipelineRunFetcher has methods to fetch PipelineRun objects from Kubernetes
+type PipelineRunFetcher interface {
+	pipelineRunNameFetcher
+	ByKey(key string) (*api.PipelineRun, error)
+}
+
+type pipelineRunNameFetcher interface {
+	ByName(namespace, name string) (*api.PipelineRun, error)
+}
```
Comment (anbrsap) [inline pkg/k8s/pipelineRunFetcher.go]: Better make two public single-method interfaces and an umbrella interface just combining the two.

File: pkg/k8s/pipelineRunFetcher.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+type pipelineRunListerFetcher struct {
```
Comment (anbrsap) [inline pkg/k8s/pipelineRunFetcher.go]: No tests exist for this type.

File: pkg/k8s/pipelineRunFetcher.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+func (f *pipelineRunListerFetcher) ByKey(key string) (*api.PipelineRun, error) {
+	return byKey(f, key)
+}
+
+type pipelineRunFetcher struct {
+	factory ClientFactory
+}
+
+// NewPipelineRunFetcher returns an operative implementation of PipelineRunFetcher
+func NewPipelineRunFetcher(factory ClientFactory) PipelineRunFetcher {
```
Comment (anbrsap) [inline pkg/k8s/pipelineRunFetcher.go]: - use a more specific name
- take only the client

```suggestion
func NewClientBasedPipelineRunFetcher(client *stewardv1alpha1.StewardV1alpha1Interface) PipelineRunFetcher {
```

File: pkg/runctl/controller.go [hunk 5]
```diff
@@ -163,16 +162,15 @@ func (c *Controller) createRunManager(pipelineRun k8s.PipelineRun) RunManager {
 // converge the two. It then updates the Status block of the Foo resource
 // with the current status of the resource.
 func (c *Controller) syncHandler(key string) error {
-	pipelineRun, err := c.pipelineRunFetcher.ByKey(key)
+	pipelineRunAPI, err := c.pipelineRunFetcher.ByKey(key)
```
Comment (anbrsap) [inline pkg/runctl/controller.go]:
```suggestion
	pipelineRunAPIObj, err := c.pipelineRunFetcher.ByKey(key)
```

File: pkg/runctl/controller_test.go [hunk 6]
```diff
@@ -135,8 +135,8 @@ func Test_Controller_syncHandler_givesUp_onPipelineRunNotFound(t *testing.T) {
 		Return(nil, nil)
 
 	// EXERCISE
-	examinee := NewController(cf, mockPipelineRunFetcher, metrics.NewMetrics())
-
+	examinee := NewController(cf, metrics.NewMetrics())
+	examinee.pipelineRunFetcher = mockPipelineRunFetcher
```
Comment (anbrsap) [inline pkg/runctl/controller_test.go]: This is still setup, not exercise.

File: pkg/runctl/controller_test.go [hunk 7]
```diff
@@ -135,8 +135,8 @@ func Test_Controller_syncHandler_givesUp_onPipelineRunNotFound(t *testing.T) {
...
 	// VERIFY
 	assert.NilError(t, examinee.syncHandler("foo/bar"))
```
Comment (anbrsap) [inline pkg/runctl/controller_test.go]: The call to `syncHandler` is the actual exercise. The `NilError` check is verification. Please separate those.

File: pkg/runctl/controller_test.go [hunk 8]
```diff
@@ -154,7 +154,8 @@ func Test_Controller_syncHandler_initiatesRetrying_on500DuringPipelineRunFetch(t
 		Return(nil, k8serrors.NewInternalError(fmt.Errorf(message)))
 
 	// EXERCISE
-	examinee := NewController(cf, mockPipelineRunFetcher, metrics.NewMetrics())
+	examinee := NewController(cf, metrics.NewMetrics())
+	examinee.pipelineRunFetcher = mockPipelineRunFetcher
 
 	// VERIFY
 	assert.ErrorContains(t, examinee.syncHandler("foo/bar"), message)
```
Comment (anbrsap) [inline pkg/runctl/controller_test.go]: dito

File: pkg/runctl/controller_test.go [hunk 9]
```diff
@@ -269,8 +270,9 @@ func resource(resource string) schema.GroupResource {
 // Return nil if not found.
 func getPipelineRun(name string, namespace string, cf *fake.ClientFactory) k8s.PipelineRun {
 	key := fake.ObjectKey(name, namespace)
-	pipelineRun, _ := k8s.NewPipelineRunFetcher(cf).ByKey(key)
-	return pipelineRun
+	fetcher := k8s.NewPipelineRunFetcher(cf)
+	pipelineRun, _ := fetcher.ByKey(key)
```
Comment (anbrsap) [inline pkg/runctl/controller_test.go]: Do not ignore errors.

File: pkg/k8s/pipelineRunFetcher.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	return byKey(f, key)
+}
+
+type clientBasedPipelineRunFetcher struct {
+	client stewardv1alpha1.StewardV1alpha1Interface
+}
+
+// NewClientBasedPipelineRunFetcher returns a PipelineRunFetcher that retrieves
+// the objects from the given API client.
+func NewClientBasedPipelineRunFetcher(client stewardv1alpha1.StewardV1alpha1Interface) PipelineRunFetcher {
```
Comment (alxsap) [inline pkg/k8s/pipelineRunFetcher.go]: Is the ClientBasedPipelineRunFetcher used in any productive code? Do we need it?

Comment (rinckm) [reply pkg/k8s/pipelineRunFetcher.go]: We currently need it for testing.

Comment (alxsap) [reply pkg/k8s/pipelineRunFetcher.go]: Could we move the code to a `_test` file?

Comment (alxsap) [reply pkg/k8s/pipelineRunFetcher.go]: Discussed with @anbrsap - keep it here.

**Pull request merged**